adapters/kidoz/kidoz.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type KidozAdapter struct {
+	endpoint string
+}
+
+func (a *KidozAdapter) MakeRequests(request *openrtb.BidRequest, _ *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	impressions := request.Imp
+	result := make([]*adapters.RequestData, 0, len(impressions))
+	errs := make([]error, 0, len(impressions))
+
+	for i, impression := range impressions {
+		if impression.Banner == nil && impression.Video == nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: "Kidoz only supports banner or video ads",
+			})
+			continue
+		}
+
+		if impression.Banner != nil {
+			banner := impression.Banner
+			if banner.Format == nil {
+				errs = append(errs, &errortypes.BadInput{
+					Message: "banner format required",
+				})
+				continue
+			}
+			if len(banner.Format) == 0 {
+				errs = append(errs, &errortypes.BadInput{
+					Message: "banner format array is empty",
+				})
+				continue
+			}
+		}
+
+		if len(impression.Ext) == 0 {
+			errs = append(errs, errors.New("impression extensions required"))
+			continue
+		}
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(impression.Ext, &bidderExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if len(bidderExt.Bidder) == 0 {
+			errs = append(errs, errors.New("bidder required"))
+			continue
+		}
+		var impressionExt openrtb_ext.ExtImpKidoz
+		err = json.Unmarshal(bidderExt.Bidder, &impressionExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if impressionExt.AccessToken == "" {
+			errs = append(errs, errors.New("Kidoz access_token required"))
+			continue
+		}
+		if impressionExt.PublisherID == "" {
+			errs = append(errs, errors.New("Kidoz publisher_id required"))
+			continue
+		}
+
+		request.Imp = impressions[i : i+1]
+		body, err := json.Marshal(request)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		result = append(result, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     a.endpoint,
+			Body:    body,
+			Headers: headers,
+		})
+	}
+
+	request.Imp = impressions
+
+	if len(result) == 0 {
+		return nil, errs
+	}
+	return result, errs
+}
+
+func (a *KidozAdapter) MakeBids(request *openrtb.BidRequest, _ *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		fallthrough
+	case http.StatusServiceUnavailable:
+		return nil, nil
+
+	case http.StatusBadRequest:
+		fallthrough
+	case http.StatusUnauthorized:
+		fallthrough
+	case http.StatusForbidden:
+		return nil, []error{&errortypes.BadInput{
+			Message: "unexpected status code: " + strconv.Itoa(responseData.StatusCode),
```

Comment:
Consider using different text in the error messages between BadInput and BadServerResponse to make the case more obvious to callers.